### DISCUSSION TITLE
Plugins order

### DISCUSSION THREAD

**Discussion by alex00888** [nuxt/nuxt#16653] [Questions]
I need to make a plugin that will use the state. And the state is created and filled by default values in another plugin. How to be sure that first plugin will always run after second and it can use the state?

**Comment by wobsoriano** [comment]
How about importing them in nuxt config?

```ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
    plugins: [
        '~/plugins/first.ts',
        '~/plugins/second.ts',
        '~/plugins/third.ts'
    ]
})
```

**Reply by alex00888** [reply]
Documentation says I should not do this:

>  All plugins in your plugins/ directory are auto-registered, so you should not add them to your nuxt.config separately.

https://v3.nuxtjs.org/docs/directory-structure/plugins

And there's nothing about 'plugins' section in the nuxt.config docs: https://v3.nuxtjs.org/docs/directory-structure/nuxt.config

**Reply by wobsoriano** [reply]
> Documentation says I should not do this:
> 
> >  All plugins in your plugins/ directory are auto-registered, so you should not add them to your nuxt.config separately.
> 
> https://v3.nuxtjs.org/docs/directory-structure/plugins
> 
> And there's nothing about 'plugins' section in the nuxt.config docs: https://v3.nuxtjs.org/docs/directory-structure/nuxt.config

You can try 😉

**Reply by kissu** [reply]
Not sure if it was removed but this is indeed following the order they are written in.

**Reply by wobsoriano** [reply]
> Not sure if it was removed but this is indeed following the order they are written in.

Still there 

https://github.com/nuxt/framework/blob/main/packages/kit/src/config/schema/build.ts#L278
https://github.com/nuxt/framework/blob/main/packages/kit/src/config/schema/_app.ts#L155

**Reply by kissu** [reply]
I meant in the documentation. I didn't dived it more than that tho.

**Reply by wobsoriano** [reply]
There is also an `extendPlugins` in the Nuxt 2 docs which lets you  change the plugins order https://nuxtjs.org/docs/directory-structure/plugins/#the-extendplugins-property

It's still available in Nuxt 3.

**Comment by aaronransley** [comment]
@alex00888 I'm not positive since I'm just scanning the source, but it looks like you may be able to mutate the `nuxt.options.plugins` array from within your module. 

Here is the source for the `addPlugin` method from the modules API; I imagine your module code could access `useNuxt().options.plugins`, and perhaps re-sort the array so that your plugin exists at a specific index.

https://github.com/nuxt/framework/blob/09d31b4aeb41c03367b7be038c8ee43fc3195e50/packages/kit/src/module/utils.ts#L111-L124

**Comment by wobsoriano** [comment]
In Nuxt 2, there's the [extendPlugins](https://nuxtjs.org/docs/directory-structure/plugins/#the-extendplugins-property) property that lets you change the plugins order created by Nuxt. You can still use it in Nuxt 3.

```ts
export default defineNuxtConfig({
  extendPlugins(plugins) {
    const pluginIndex = plugins.findIndex(
      ({ src }) => src === '~/plugins/shouldBeFirst.ts'
    )
    const shouldBeFirstPlugin = plugins[pluginIndex]

    plugins.splice(pluginIndex, 1)
    plugins.unshift(shouldBeFirstPlugin)

    return plugins
  }
})
```

**Comment by jawngee** [comment]
`extendPlugins` doesn't seem to work as the function never seems to be called.  We ended up having to do this with a custom module:

In `~/modules/plugin-loader.ts`:
```js
import {ModuleOptions} from "@nuxt/schema";

export default function PluginLoaderModule(moduleOptions:ModuleOptions) {
    if (this.nuxt.options.otherPlugins) {
        this.nuxt.options.otherPlugins.forEach(plugin => this.nuxt.options.plugins.push({ src: plugin }))
    }
}
```

In `nuxt.config.ts`:
```js
export default defineNuxtConfig({
    ssr: false,
    target: 'static',

    otherPlugins: [
        '~/other-plugins/axios-plugin.ts',
        '~/other-plugins/auth-plugin.ts',
    ],

    modules: [
        '~/modules/plugin-loader.ts'
    ],
});
```

In `index.d.ts`:
```js
declare module '@nuxt/schema' {
  interface NuxtConfig {
    otherPlugins?: string[]
  }
}

export { }
```

And then instead of dropping plugins in `~/plugins/` I drop them in `~/other-plugins/` and make sure to add them to the `otherPlugins` property in the config.

**Comment by Jlaird** [comment]
Has anyone found a way to disable the auto import of plugins? We too are having the issue where they are not imported in the correct order breaking our app.

**Reply by multiplehats** [reply]
There's an option in Nuxt config `autoimports` but it's only for composables :(

**Comment by valqelyan** [comment]
Any updates ?

**Comment by ironexdev** [comment]
It looks like plugins are loaded in alphabetical order based on their (file) names - prefixing them with numbers should be enough to control the order

** Plugin files** (each file contains console.log <number> <type>, type is based on the file name "client", "server", "unspecified"
```
1.client.ts
export default defineNuxtPlugin((nuxtApp) => {
  console.log(1 + " client")
})

2.ts
export default defineNuxtPlugin((nuxtApp) => {
  console.log(2 + " unspecified")
})

3.server.ts
export default defineNuxtPlugin((nuxtApp) => {
  console.log(3 + " server")
})

4.ts
export default defineNuxtPlugin((nuxtApp) => {
  console.log(4 + " unspecified")
})

5.server.ts
export default defineNuxtPlugin((nuxtApp) => {
  console.log(5 + " server")
})

6.client.ts
export default defineNuxtPlugin((nuxtApp) => {
  console.log(6 + " client")
})
```


**Server-side** console output
<img width="141" alt="image" src="https://user-images.githubusercontent.com/24256329/192116276-0a711305-bcba-41f7-8b47-6434e525abfb.png">

**Client-side** console output
<img width="126" alt="image" src="https://user-images.githubusercontent.com/24256329/192116283-674f2157-47a4-4e28-a1ac-afe7aff26707.png">

**Comment by miketromba** [comment]
Controlling the order of plugin registration is a common requirement, so I think this should be added as an explicit recommendation in the docs. 

Right now, it just says "Plugins are called in order sequentially and before everything else" in a small part of the docs. I don't think this is verbose enough. A section titled something like "Plugin registration order" would be very helpful.

Example: 
## Plugin registration order
Sometimes, you may have a plugin that depends on another plugin. In this case, it's important to ensure that the first plugin is registered before the one that depends on it. To do this, you can prefix a number to the plugin's file name like so:
- plugins/1.foo.ts (Loaded first)
- plugins/2.bar.ts (Loaded second... it can access anything that was injected by 1.foo.ts)

**Reply by miketromba** [reply]
PR https://github.com/nuxt/framework/pull/8096

**Comment by H-M-H-cmd** [comment]
For me the order of numbers doesn't work well, I just copied the code that I depend on from the required plugin and past it in the plugin that must start after the first plugin

**Comment by jawngee** [comment]
If you are using nuxt 3, then you can actually define plugin dependencies:  https://nuxt.com/docs/guide/directory-structure/plugins#plugins-with-dependencies

**Reply by H-M-H-cmd** [reply]
thanks I will try it

**Comment by TheAlexLichter** [comment]
With Nuxt 3's [plugin object syntax](https://www.youtube.com/watch?v=2aXZyXB1QGQ), you can define explicit dependencies (say plugin Z depends on A) :)

**Reply by H-M-H-cmd** [reply]
I was actually watching your video now, so cool ;)
